### PR TITLE
refactor!: migrate to typescript

### PR DESCRIPTION
**Pull request by santimirandarp**
Breaking change is due to a reworded error message. [From this line.](https://github.com/mljs/regression-base/pull/15/commits/78ec8ccd226513db767ec3232fc7826fa22d029b#diff-bfe9874d239014961b1ae4e89875a6155667db834a410aaaa2ebe3cf89820556L50).

It's documented in each commit message.

### CODE REVIEW

File: .github/workflows/typedoc.yml [hunk 1]
```diff
@@ -0,0 +1,32 @@
...
+env:
+  NODE_VERSION: 16.x
+  ENTRY_FILE: 'src/index.ts'
+
+jobs:
+  deploy:
+    runs-on: ubuntu-latest
+    steps:
+      - uses: actions/checkout@v2
+      - uses: actions/setup-node@v2
```
Comment (targos) [inline .github/workflows/typedoc.yml:17]: These do not correspond to the latest template.

Comment (santimirandarp) [reply .github/workflows/typedoc.yml:17]: Thanks.

Script taking this from cheminfo `.github` repository I think. 

Maybe we can update it there, and also here.

Comment (targos) [reply .github/workflows/typedoc.yml:17]: Looks up to date [in cheminfo](https://github.com/cheminfo/.github/blob/main/workflow-templates/typedoc.yml#L16) too.

**Comment by santimirandarp** [comment]
I assume `NumberArray` is a better fit than `AnyArray` (which could be imported) but it may need to be modified.

File: src/index.ts [hunk 2]
```diff
@@ -1,55 +1,65 @@
...
   constructor() {
     if (new.target === BaseRegression) {
       throw new Error('BaseRegression must be subclassed');
     }
   }
 
-  predict(x) {
+  predict(x: number): number;
+  predict(x: number[]): number[];
+  predict(x: number | number[]) {
```
Comment (targos) [inline src/index.ts:22]: why not NumberArray here?

**Comment by lpatiny** [comment]
I think you could use: https://github.com/mljs/spectra-processing/blob/main/src/xy/xyCheck.ts

File: src/index.ts (hunk 2 above)
Comment (lpatiny) [reply src/index.ts:22]: NumberArray is in cheminto-types: https://github.com/cheminfo/cheminfo-types/blob/main/src/core/NumberArray.d.ts
This package (cheminfo-types) is where we place also the share types definition.

**Comment by santimirandarp** [comment]
> I think you could use: https://github.com/mljs/spectra-processing/blob/main/src/xy/xyCheck.ts

They do almost the same. I'm not sure about the change since the current function is also exported in `index.ts` and it's the input types are different.

**Comment by santimirandarp** [comment]
Updated branch name. Seems it still needs the [branch update in codecov]( https://app.codecov.io/gh/mljs/regression-base)

**Comment by targos** [comment]
> > I think you could use: https://github.com/mljs/spectra-processing/blob/main/src/xy/xyCheck.ts
> 
> It'd be useful. But I'm not sure about the change since the current function is also exported in `index.ts` and the input types are different.

It would be overkill. This package has basically no dependencies right now. `spectra-processing` is a huge one.

**Comment by targos** [comment]
> Updated branch name in github actions. Seems it still needs [repointing the branch to main in codecov](https://app.codecov.io/gh/mljs/regression-base) to show the correct coverage in future changes.

done

File: package.json [hunk 3]
```diff
@@ -36,14 +40,17 @@
...
   "devDependencies": {
-    "@babel/plugin-transform-modules-commonjs": "^7.16.8",
-    "eslint": "^8.10.0",
-    "eslint-config-cheminfo": "^7.2.2",
-    "jest": "^27.5.1",
-    "prettier": "^2.5.1",
-    "rollup": "^2.68.0"
+    "@types/jest": "^29.5.3",
+    "@vitest/coverage-v8": "^0.34.1",
+    "cheminfo-types": "^1.7.2",
```
Comment (targos) [inline package.json:48]: You're using this package in the published types. It must be in prod dependencies

File: package.json [hunk 4]
```diff
@@ -32,18 +36,17 @@
...
+    "@vitest/coverage-v8": "^0.34.1",
+    "eslint": "^8.46.0",
+    "eslint-config-cheminfo-typescript": "^12.0.4",
+    "prettier": "^3.0.1",
+    "rimraf": "^5.0.1",
+    "typescript": "^5.1.6",
+    "vitest": "^0.34.1"
   }
-}
+}
```
Comment (targos) [inline package.json:52]:
```suggestion
}

```

**Review by targos — APPROVED** [review]

**Pull request merged**